src/vmm/src/gdb/server.rs

```diff
@@ -0,0 +1,101 @@
+};
+use kvm_ioctls::VcpuFd;
+use vm_memory::GuestAddress;
+
+use super::event_loop::event_loop;
+use super::target::Error;
+use crate::logger::info;
+use crate::{Vcpu, Vmm};
+
+const KVM_DEBUG_ENABLE: u64 = 0x0600;
+
+fn set_kvm_debug(control: u32, vcpu: &VcpuFd, addrs: &[GuestAddress]) -> Result<(), Error> {
+    let mut dbg = kvm_guest_debug {
+        control,
+        ..Default::default()
+    };
+
+    dbg.arch.debugreg[7] = KVM_DEBUG_ENABLE;
+
+    for (i, addr) in addrs.iter().enumerate() {
+        dbg.arch.debugreg[i] = addr.0;
+        // Set global breakpoint enable flag for the specific breakpoint number by setting the bit
+        dbg.arch.debugreg[7] |= 2 << (i * 2);
+    }
+
+    vcpu.set_guest_debug(&dbg).map_err(|_| Error::VcpuKvmError)
+}
+
+/// Configures the VCPU for debugging and sets the hardware breakpoints on the vcpu
+pub fn vcpu_set_debug(vcpu: &VcpuFd, addrs: &[GuestAddress], step: bool) -> Result<(), Error> {
+    let mut control = KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_HW_BP | KVM_GUESTDBG_USE_SW_BP;
+    if step {
+        control |= KVM_GUESTDBG_SINGLESTEP;
+    }
+
+    set_kvm_debug(control, vcpu, addrs)
+}
+
+/// Injects a BP back into the guest kernel for it to handle, this is particularly useful for the
+/// kernels selftesting which can happen during boot.
+pub fn vcpu_inject_bp(vcpu: &VcpuFd, addrs: &[GuestAddress], step: bool) -> Result<(), Error> {
+    let mut control = KVM_GUESTDBG_ENABLE
+        | KVM_GUESTDBG_USE_HW_BP
+        | KVM_GUESTDBG_USE_SW_BP
+        | KVM_GUESTDBG_INJECT_BP;
+
+    if step {
+        control |= KVM_GUESTDBG_SINGLESTEP;
+    }
+
+    set_kvm_debug(control, vcpu, addrs)
+}
+
+/// This method will kickstart the GDB debugging process, it takes in the VMM object, a slice of
+/// the paused Vcpu's, the GDB event queue which is used as a mechanism for the VCPU's to notify
+/// our GDB thread that they've been paused, then finally the entry address of the kernel.
+///
+/// Firstly the function will start by configuring the vcpus with KVM for debugging
+///
+/// This will then create the GDB socket which will be used for communication to the GDB process.
+/// After creating this, the function will block while waiting for GDB to connect.
+///
+/// After the connection has been established the function will start a new thread for handling
+/// communcation to the GDB server
+pub fn gdb_thread(
+    vmm: Arc<Mutex<Vmm>>,
+    vcpus: &[Vcpu],
+    gdb_event_receiver: Receiver<usize>,
+    entry_addr: GuestAddress,
+) -> Result<(), Error> {
+    // We register a hw breakpoint at the entry point as GDB expects the application
+    // to be stopped as it connects. This also allows us to set breakpoints before kernel starts
+    vcpu_set_debug(&vcpus[0].kvm_vcpu.fd, &[entry_addr], false)?;
+
+    for vcpu in vcpus.iter().skip(1) {
+        vcpu_set_debug(&vcpu.kvm_vcpu.fd, &[], false)?;
+    }
+
+    let path = Path::new("/tmp/gdb.socket");
```

Comment:
This should either be configurable (maybe with config file only?) or it should be relative to the firecracker `cwd`. Otherwise who knows is `/tmp/gdb.socket` is already taken by smth. Or if you need to debug 2 firecracker processes at the same time.